/**
  * Some common messages.
  * Version : 1
  * All string are encoded in UTF-8.
  */

package Protos.Common;

message Null {
}

message Error {
   optional string message = 1;
}

message Language {
   required string lang = 1; // ISO-639 (two letters)
   optional string country = 2; // ISO-3166
}

// For identify a chunk or a user.
message Hash {
   optional bytes hash = 1; // 20 bytes. If it doesn't exist the hash is null.
}

message IP {
   enum Type {
      IPv4 = 0;
      IPv6 = 1;
   }
   required Type type = 1;
   required bytes ip = 2; // type == IPv4: 4 bytes. type == IPv6: 16 bytes. Big endian.
}

// A shared directory matches an existing handle, thus there is no need to have the complete path to it.
message SharedDir {
   required Hash id = 1; // Identify a shared directory.
   optional string shared_name = 2; // A name to identify the shared directory by human. It may correspond to the name of the shared directory itself. It may be used when creating the directories of a new downloading file.
}

// Used for a directory or a file.
message Entry {
   enum Type {
      FILE = 0;
      DIR = 1;
   }
   required Type type = 1;
   required string path = 2; // Relative path from the sharing directory. For example : "/Videos/My cat/". A root directory has a one slash : "/". A shared directory has an empty path.
   required string name = 3; // Directory name or file name. For example : "MyLOLCat.avi". A root directory has an undetermined name.
   required uint64 size = 4; // [bytes].
   
   optional SharedDir shared_dir = 5; // The shared directory of the entry.
   optional bool exists = 6 [default = true]; // An entry may not yet be created.
   optional bool is_empty = 7 [default = true]; // Only for a Dir. Warning: a directory can be not empty and has a size of 0.
   
   repeated Hash chunk = 8; // Only for a file. No chunk means that the peer doesn't know the hashes.
}

message Entries
{
   repeated Common.Entry entry = 1;
}

// A result following a search.
// Entries may not be sorted in any particular way.
// See 'Network.Find' and 'GUI.GUINetwork' for more information.
message FindResult {
   message EntryLevel {
      required uint32 level = 1;
      required Entry entry = 2; // The entry must have the field 'shared_dir'.
   }
   required uint64 tag = 1;
   repeated EntryLevel entry = 2;
   optional Common.Hash peer_id = 3;
}

message Interface {
   message Address {
      enum Protocol {
         IPv4 = 1;
         IPv6 = 2;
      }
      required string address = 1;
      required Protocol protocol = 2;
      required bool listened = 3;
   }
   required uint32 id = 1;
   required string name = 2;
   repeated Address address = 3;
   optional bool isUp = 4;
}
